Add option for channel 'whitelisting' (solves #3)

mrc0mmand opened the pull request:
This patch adds an option `smartfilter_ignored_chans` which can be used for specifying channels for which the join/part/quit/nick filtering will be skipped. It should solve the issue #3.

mrc0mmand committed:
Add option for channel 'whitelisting' (solves #3)

mrc0mmand committed:
Add key check & fix garbage collector

mrc0mmand committed:
Add option 'smartfilter_filter_nick'

Irssi sends one NICK signal per server and without specifying
a current channel, thus the signal is ignored for all channels,
even for those, which are whitelisted. This commit adds a new
option, which (at least) allows to whitelist all nick changes.

mrc0mmand committed:
Ignore first signal of each nick

mrc0mmand committed:
Remove invalid channel variable from QUIT

mrc0mmand commented:
I'm not entirely sure about 0b75511. Maybe the activity should be based solely on user's messages instead of all signals (to reduce join/part noise from idle users), but that's just a matter of opinion I'd say.

lifeforms commented:
Thanks for the PR! I will check it next weekend :)

lifeforms commented:
I tested the changes and have a few questions:
1. When an inactive user quits and then rejoins the channel, the rejoin seems to be shown instead of hidden. Is this intentional? I personally prefer this to remain hidden.
2. I didn't get the ignore function to work. I tried `/set smartfilter_ignored_chans #chan1 #chan2` (and also `chan1 chan2`) but this did not seem to affect filtering. When someone joins the channel, I still don't always see it. My perl is not strong enough to see why though...

mrc0mmand commented:
Sorry for the delayed answer and thanks for checking my PR.

As for found issues:
1. This is unfortunately caused by adding users into 'active' list even when they join/part. As I said before, solution would be to ignore join/part messages when the user is not on current 'active' list.
2. Uh, that's strange, I've been using this modified version for about a month and haven't noticed this behavior so far. I guess that would need some debugging.

Also, there's a problem with filtering NICK and QUIT messages, because signal for them is sent only once per server, so if you filter them for one channel, you'll eventually filter them for all channels on that server. One workaround would be the change I commited in c5011f0 for NICK and it would need similar one for QUIT, but that skips NICK/QUIT filtering for all channels. Alternative workaround I can think of would be parsing and filtering `gui print text` signal, but that's kind of inconvenient, because the signal could be called multiple times for one message, so the message would end up split into multiple messages.

mrc0mmand committed:
Fix issues discussed in PR #5

- Ignore function should now work as expected
- Base user activity solely on their 'message activity'
    - This should solve the join/part spam
- Initialize ignore_chans array only once on a script load
- Add an option to ignore filtering of a QUIT signal
    - Behavior is unfortunately same as with NICK signal, so the signal
      is not filtered globally for all channels/servers

mrc0mmand commented:
After a longer delay I finally fixed most of the issues in this PR.

1) The ignore function should now work as expected. I somehow used a wrong grep syntax (and I can't believe how I didn't notice that before).
2) The join/part spam should be fixed as well, I implemented the solution mentioned in my previous comment, so the user activity is now based only on their messages (and so far it works pretty well).
3) I added a new option `smartfilter_filter_quit` which enables/disables filtering of QUIT signal. Unfortunately, it has the same problem as the NICK one - the signal is (not) filtered for all servers. I'm still not sure about the NICK/QUIT filtering, maybe I'm missing the easiest solution.

mrc0mmand committed:
Filter NICK and QUIT signals separately for each channel

mrc0mmand commented:
Well, after further investigation I found a way how to parse NICK and QUIT messages for each channel separately, thus options `smartfilter_filter_quit` and `smartfilter_filter_nick` won't be needed anymore. I did several manual tests and it seems to work, so I hope it will stay this way.

lifeforms commented:
Hey, sorry for the wait, I finally got around to testing your patch, it seems to perform okay! :) 

One thing is, when i change ignored channels, the effect doesn't seem to take until next restart. I think it's due to these lines being in main, so they are only executed on plugin initialization:

```
my $ign_chans = Irssi::settings_get_str('smartfilter_ignored_chans');
@ignored_chans = split /\s+/, $ign_chans;
```

What do you think about moving those into the `checkactive` subroutine? If it's for performance reasons, maybe we could also cache `$ign_chans` and only do the regexp if the setting has changed.

If you can add explanation to the README for the new options, that would be awesome, also feel free to add your name to the contributors. I can do this too if you'd like me to do it.

mrc0mmand committed:
Reload ignored channels on settings change

mrc0mmand committed:
Add smartfilter_ignored_chans option into README.md

mrc0mmand commented:
Recreating the `@ignored_chans` array on settings change definitely makes sense. Thankfully, for this purpose, there's a `setup changed` signal, thus I created a new subroutine `smartfilter_settings` and hooked it to the signal.

I also added a `smartfilter_ignored_chans` option into README.md with a brief description, hopefully it's all right.

lifeforms commented:
Your pull request is now 100% awesome. I merged it! I'm sure lots of people have been waiting for this, so lots of thanks for the PR and sticking with it throughout all my nagging. :)
